types/opt/value.go

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+package opt
+
+import (
+	"fmt"
+	"reflect"
+
+	jsonv2 "github.com/go-json-experiment/json"
+	"github.com/go-json-experiment/json/jsontext"
+)
+
+// Value is an optional value to be JSON-encoded.
+// With [encoding/json], a zero Value is marshaled as a JSON null.
+// With [github.com/go-json-experiment/json], a zero Value is omitted from the
+// JSON object if the Go struct field specified with omitzero.
+// The omitempty tag option should never be used with Value fields.
+type Value[T any] struct {
+	value T
+	set   bool
+}
+
+// Equal reports whether the receiver and the other value are equal.
+// If the template type T in Value[T] implements an Equal method, it will be used
+// instead of the == operator for comparing values.
+type equatable[T any] interface {
+	// Equal reports whether the receiver and the other values are equal.
+	Equal(other T) bool
+}
+
+// ValueOf returns an optional Value containing the specified value.
+// It treats nil slices and maps as empty slices and maps.
+func ValueOf[T any](v T) Value[T] {
+	return Value[T]{value: v, set: true}
+}
+
+// String implements [fmt.Stringer].
+func (o *Value[T]) String() string {
+	if !o.set {
+		return fmt.Sprintf("(empty[%T])", o.value)
+	}
+	return fmt.Sprint(o.value)
+}
+
+// Set assigns the specified value to the optional value o.
+func (o *Value[T]) Set(v T) {
+	*o = ValueOf(v)
+}
+
+// Clear resets o to an empty state.
+func (o *Value[T]) Clear() {
+	*o = Value[T]{}
+}
+
+// IsSet reports whether o has a value set.
+func (o *Value[T]) IsSet() bool {
+	return o.set
+}
+
+// Get returns the value of o, panicking if it hasn't been set.
+func (o Value[T]) Get() T {
+	if !o.set {
+		panic(fmt.Sprintf("%v is not set", reflect.TypeFor[T]()))
```

Comment:
Probably want to just return the zero value of `T` in a similar way to how indexing into a non-existent map entry (e.g., `myMap[noExistKey]`) returns the zero element type instead of panicking..

Reply:
Done.